Editorial: add "add an event listener" hook

annevk opened the pull request:
This hook ensures that any special casing in addEventListener() is shared with event handler attributes.

Helps with https://github.com/w3c/ServiceWorker/issues/1004.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/dom/596.html" title="Last updated on Mar 14, 2018, 8:35 AM GMT (4ebd185)">Preview</a> | <a href="https://whatpr.org/dom/596/8e754b5...4ebd185.html" title="Last updated on Mar 14, 2018, 8:35 AM GMT (4ebd185)">Diff</a>

domenic commented:
Since event handlers don't have options, maybe this should accept a few booleans instead of a dictionary-or-boolean (with good defaults that make sense for event handlers)?

annevk commented:
I'd rather not make this an ever expanding set of arguments. I'd be okay with making the last argument optional to make the HTML caller slightly less ugly, but I'm not sure it's worth it.

domenic commented:
The issue is that for anyone that calls it except for the addEventListener() definition, they have to manually assemble an AddEventListenerOptions dict, which feels like it should be specific to the addEventListener API and to JS consumers of it.

annevk commented:
I guess you're considering a third caller we haven't decided on adding thus far. I guess if that might not use AddEventListenerOptions we should make that argument optional for now and not pass it from HTML. Then whenever this third caller comes around we can restructure this without having to change HTML.

domenic commented:
Yeah I'm kind of thinking of https://domenic.github.io/async-local-storage/#add-a-simple-event-listener

annevk commented:
Hmm, I suppose we could use that signature, requiring the caller to create an event listener concept wrapper.

domenic commented on dom.bs:
```diff
@@ -328,7 +328,7 @@ method, passing the same arguments.
 <var ignore>ev</var> is the <a>event</a>. It is
 passed as argument to
-<a>event listener</a>'s <b>callback</b>
+<a>event listener</a>'s <a for="event listener">callback</a>
```
Yesss I've been wanting this for a while.

domenic commented on dom.bs:
```diff
@@ -958,48 +958,48 @@ are not to be used for anything else. [[!HTML]]
-  a boolean, in which case the method behaves exactly as if the value was specified as
+  <p>The <var>options</var> argument sets listener-specific options. For compatibility this can be a
+  boolean, in which case the method behaves exactly as if the value was specified as
   <var>options</var>' <code>capture</code> member.
```
While here you could make all these `<code>`s link to the dictionary member definitions, as is done in the algorithms below.

domenic commented on dom.bs:
```diff
@@ -958,48 +958,48 @@ are not to be used for anything else. [[!HTML]]
-  <a>event</a>'s {{Event/eventPhase}} attribute value is {{Event/CAPTURING_PHASE}}. Either way,
-  <b>callback</b> will be invoked if <a>event</a>'s {{Event/eventPhase}} attribute value is
-  {{Event/AT_TARGET}}.
+  <p>When set to true, <var>options</var>' <code>capture</code> member prevents
```
While here you could consistify to use "options's" per https://wiki.whatwg.org/wiki/Style#Grammar

annevk replied:
I thought there was still an exception when the word is plural?

domenic replied:
We've had this discussion before :). `options` is not a class of values; it is a single JS value. So it's not really plural in the grammatical sense that would trigger leaving off the s, where `capture` belongs to a collection of options.

Stated another way, here "options" is short for "options variable", which is not plural.

A case where it would be plural is "the stock options' combined value was a billion dollars."

domenic commented on dom.bs:
```diff
@@ -1038,13 +1038,12 @@ must return a new {{EventTarget}}.
-<p>The
-<dfn method for=EventTarget><code>addEventListener(<var>type</var>, <var>callback</var>, <var>options</var>)</code></dfn>
-method, when invoked, must run these steps:
+<p>To <dfn export>add an event listener</dfn> given an <code>EventTarget</code> object
```
`{{EventTarget}}` instead of just code

domenic commented on dom.bs:
```diff
@@ -1054,34 +1053,53 @@ method, when invoked, must run these steps:
+</ol>
+
+<p class=note>The <a>add an event listener</a> concept exists to ensure
+<a>event handler attributes</a> use the same code path. [[HTML]]
```
This link is not working

annevk commented:
I think I addressed the feedback, but I ended up touching some unrelated sections since it seemed good to make those suggested editorial changes consistently throughout.
